rust/mdb_shard/src/shard_file_manager.rs

```diff
@@ -129,23 +164,76 @@ impl ShardFileManager {
         new_shards: &[MDBShardFile],
         shards_are_permanent: bool,
     ) -> Result<()> {
-        let mut current_lookup = self.shard_file_lookup.write().await;
-
-        for s in new_shards {
+        let mut shards_lg = self.shard_file_lookup.write().await;
+
+        // Go through and register the shards in order of newest to oldest
+        let mut new_shards: Vec<(&MDBShardFile, SystemTime)> = new_shards
+            .iter()
+            .map(|s| {
+                let modified = std::fs::metadata(&s.path)?.modified()?;
+                Ok((s, modified))
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        // Compare in reverse order to sort from newest to oldest
+        new_shards.sort_by(|(_, t1), (_, t2)| t2.cmp(t1));
+        let num_shards = new_shards.len();
+
+        for (s, _) in new_shards {
             debug!(
                 "register_shards: Registering shard {:?} at {:?}.",
                 s.shard_hash, s.path
             );
-            current_lookup
-                .entry(s.shard_hash)
-                .or_insert((s.clone(), shards_are_permanent));
+            let cur_index = shards_lg.shard_list.len();
+            let mut inserted = false;
+
+            shards_lg.shards.entry(s.shard_hash).or_insert_with(|| {
+                inserted = true;
+                cur_index
+            });
+
+            if inserted {
+                shards_lg.shard_list.push((s.clone(), shards_are_permanent));
+                if cur_index < u16::MAX as usize
+                    && shards_lg.chunk_lookup.len() + s.shard.total_num_chunks()
+                        < shards_lg.chunk_index_max_size
+                {
+                    for (h, (cas_start_index, cas_chunk_offset)) in s.read_all_truncated_hashes()? {
+                        if cas_chunk_offset > u16::MAX as u32 {
+                            break;
+                        }
+                        let cas_chunk_offset = cas_chunk_offset as u16;
```

Comment:
Is this casting safe?

Reply:
Given the test above, yes.  Longer ones are ignored.